### PR TITLE
`Communication`: Show messaging page if only communication is enabled

### PR DESCRIPTION
**Pull request by lennart-keller**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).(https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

We are in the process of unifying the communication and messaging pages of Artemis. Even if only communication is enabled, the messaging page should be used to create new messages.

### Description

If communication is enabled, also the messaging page will be activated in the course. However, it only shows course-wide channels. The client also hides buttons to create new channels, group chats, and one-to-one chats and the server only allows to create these if messaging is enabled.

### Steps for Testing

Prerequisites:
- 1 Instructor (to change course settings)
- 1 Student

1. Deactivate messaging and communication in a course
    1. Neither the communication tab or messaging tab should be visible
    2. On a lecture/exercise details page, there is no "communication" sidebar to the right
2. Activate communication
    1. Both the communication tab and messaging tab should be visible
    2. The messaging tab should only show course-wide channels (no group chats etc.)
    3. Test all the buttons you see with both the instructor and students (hide, favorite, settings, "go to exercise", ...)
    4. Write some messages and check this works on lecture/exercise pages and the messages tab
    5. Write a message with a user mention: Clicking on the mention should NOT open a one-to-one chat
3. Add both communication and messaging
    1. Now, also group chats and one-to-one messages should be visible. Check you can create new ones.
4. Disable communication
    1. The communication tab is hidden, the messages tab stays the same

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [course-management-tab-bar.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5754/latest/artifact/shared/Coverage-Report-Client-Tests/app/course/manage/course-management-tab-bar/course-management-tab-bar.component.ts.html) | 93.42% | ✅ |
| [course.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5754/latest/artifact/shared/Coverage-Report-Client-Tests/app/entities/course.model.ts.html) | 100% | ✅ |
| [exam-update.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5754/latest/artifact/shared/Coverage-Report-Client-Tests/app/exam/manage/exams/exam-update.component.ts.html) | 90.07% | ✅ |
| [exercise-title-channel-name.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5754/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/shared/exercise-title-channel-name/exercise-title-channel-name.component.ts.html) | 100% | ✅ |
| [lecture-title-channel-name.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5754/latest/artifact/shared/Coverage-Report-Client-Tests/app/lecture/lecture-title-channel-name.component.ts.html) | 100% | ✅ |
| [conversation-selection-sidebar.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5754/latest/artifact/shared/Coverage-Report-Client-Tests/app/overview/course-conversations/layout/conversation-selection-sidebar/conversation-selection-sidebar.component.ts.html) | 93.37% | ✅ |
| [conversation.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5754/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/metis/conversations/conversation.service.ts.html) | 96.42% | ✅ |
| [message-reply-inline-input.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5754/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/metis/message/message-reply-inline-input/message-reply-inline-input.component.ts.html) | 78.37% | ✅ |
| [metis.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5754/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/metis/metis.service.ts.html) | 92.57% | ✅ |
| [posting-content-part.components.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5754/latest/artifact/shared/Coverage-Report-Client-Tests/app/shared/metis/posting-content/posting-content-part/posting-content-part.components.ts.html) | 94.44% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [ChannelRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5754/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository.metis.conversation/ChannelRepository.html) | 56% | ✅ |
| [AnswerMessageService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5754/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.metis/AnswerMessageService.html) | 97% | ✅ |
| [ConversationMessagingService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5754/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.metis/ConversationMessagingService.html) | 92% | ✅ |
| [ChannelService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5754/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.metis.conversation/ChannelService.html) | 94% | ✅ |
| [ConversationService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5754/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.metis.conversation/ConversationService.html) | 95% | ✅ |
| [ChannelResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5754/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.metis.conversation/ChannelResource.html) | 92% | ✅ |
| [ConversationResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5754/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.metis.conversation/ConversationResource.html) | 92% | ✅ |

### Screenshots

If only communication is enabled, the messaging tab will still be shown. However, the sidebar only shows course-wide channels and hides buttons to create new channels, group chats and one-to-one chats:
communication-only

If messaging is enabled, the sidebar shows group chats etc.:
communication-and-messaging

### CODE REVIEW

**Review by JanaNF — COMMENTED** [review]
Tried to test on legacy_ts11 and legacy_ts2 but I couldn’t connect to the server. 

I wrote with Laurenz and he said:
_A new config variable was introduced by a PR, but this is not included in the config for TS11. This is why it always fails at startup.
I think the PR was merged without regard to the config change and is now causing problems on all test servers where the config has not yet been adjusted
The PR with the config change is already in Develop and I suspect that if your PRs are running with the latest changes from Develop, then there will be problems_

I think your PR is affected.

**Review by JanaNF — APPROVED** [review]
I could test your PR on legacy_ts11 again :tada: Worked as expected.

**Review by AntonGeTUM — DISMISSED** [review]
Tested on TS11, works as described

**Review by MaximilianAnzinger — CHANGES_REQUESTED** [review]
Mostly looks good.

Left some susgestions & ArchitectureTest fails.

File: src/main/java/de/tum/in/www1/artemis/service/metis/ConversationMessagingService.java [hunk 1]
```diff
@@ -120,34 +124,49 @@ public Post createMessage(Long courseId, Post newMessage) {
         conversationParticipantRepository.updateLastReadAsync(author.getId(), conversation.getId(), ZonedDateTime.now());
 
         var createdMessage = conversationMessageRepository.save(newMessage);
+        log.debug("      conversationMessageRepository.save DONE");
```
[the remark this answers is not in the document]
Comment (lennart-keller) [reply src/main/java/de/tum/in/www1/artemis/service/metis/ConversationMessagingService.java:126]: These should stay (for now)

File: src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java [hunk 2]
```diff
@@ -334,6 +334,9 @@ public Channel updateExamChannel(Exam originalExam, Exam updatedExam) {
     }
 
     private Channel updateChannelName(Channel channel, String newChannelName) {
+        if (channel == null) {
```
[the remark this answers is not in the document]
Comment (lennart-keller) [reply src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java:337]: You are right, I moved the null check to another place, where it makes more sense.

File: src/main/webapp/app/overview/course-conversations/layout/conversation-selection-sidebar/conversation-selection-sidebar.component.html
[the remark this answers is not in the document]
Comment (lennart-keller) [reply src/main/webapp/app/overview/course-conversations/layout/conversation-selection-sidebar/conversation-selection-sidebar.component.html:1]: This already is a helper boolean, which is just set once in `ngOnInit`

File: src/main/webapp/app/course/manage/course-management-tab-bar/course-management-tab-bar.component.html [hunk 3]
```diff
@@ -27,7 +27,7 @@
             <span jhiTranslate="artemisApp.iris.settings.button.course.title">Iris</span>
             <fa-icon [icon]="faArrowUpRightFromSquare"></fa-icon>
         </a>
-        <a class="tab-link" *ngIf="isMessagingEnabled(course)" [routerLink]="['/courses', course.id, 'messages']" routerLinkActive="active">
+        <a class="tab-link" *ngIf="isMessagingEnabled(course) || isCommunicationEnabled(course)" [routerLink]="['/courses', course.id, 'messages']" routerLinkActive="active">
```
[the remark this answers is not in the document]
Comment (lennart-keller) [reply src/main/webapp/app/course/manage/course-management-tab-bar/course-management-tab-bar.component.html]: I created helper variables

**Review by MaximilianAnzinger — APPROVED** [review]
Thx for integrating my suggestions LGTM 👍

**Review by basak-akan — APPROVED** [review]
Code looks good 👍

**Review by sarpsahinalp — APPROVED** [review]
Manually tested on ts2. Seems to be working!

**Review by reschandreas — APPROVED** [review]
retested locally after demonstration in testing session two days ago, works as expected! also thanks for fixing the unrelated issues found in the testing session!

**Comment by Gusti2010** [comment]
tested manually on ts11_Legacy. Works as expected!

**Review by Gusti2010 — APPROVED** [review]
tested manually on ts11_Legacy. Works as expected!

**Review by rstief — APPROVED** [review]
tested on ts2. Everything works as expected.

**Pull request merged**